### ISSUE TITLE
indenting in nested environment off by one indent

### ISSUE THREAD

**Issue by Konfekt** [lervag/vimtex#1430]
Hitting `=ae` on with `&shiftwidth=2`
```tex
\begin{equation*}
  \xymatrix{
     A & B
  }
\end{equation*}
```
turns it into
```tex
\begin{equation*}
  \xymatrix{
    A & B
  }
  \end{equation*}
```
Expected, no change, that is:
```tex
\begin{equation*}
  \xymatrix{
    A & B
  }
\end{equation*}
```

**Comment by lervag** [comment]
The ampersand parsing that is used for aligning ampersands is kicking in and the parsing gets confused. The reason is that it is parsed backwards, and until now I thought it safe to assume that the regions where ampersands are used are all environments. Thus I could parse the "depth" by looking at `\end` ... `\begin` pairs. However, in this case, I would need to be more rigorous and parse the command delimiters as well.

This is much more challenging! I don't immediately see a solution. In the meantime, you can turn off ampersand aligning with `g:vimtex_indent_on_ampersands`.

**Comment by lervag** [comment]
Ok, I've tried to fix it. I'm afraid it will have unwanted side effects, and the fix will probably have to either be improved, or it may (probably more likely) be reverted.

**Comment by Konfekt** [comment]
Thank you very much and for pointing out `g:vimtex_indent_on_ampersands` as an alternative in case this doesn't work out.

**Comment by Konfekt** [comment]
This does not work for environments containing a backslash though: The issue persists for the example
```tex
\begin{equation*}
	\xymatrix{
	  A\a & B
	}
\end{equation*}
```

**Comment by lervag** [comment]
I think I fixed that, although I'm still not sure this will work out in the end.

**Comment by krishnakumarg1984** [comment]
@lervag, this is getting to the point wherein it is worth seriously pondering over a mechanism to provide an interface to `latexindent.pl` by automatically setting up the `equalprg` and fall back to the internal formatter if `latexindent` is not found in the PATH.  

I just faced this issue this afternoon. 

```
\begin{flushright}
    \begin{tabular}{@{}p{.4in}p{2.1in}@{}}
       Signed: & \hrulefill \\
                    & Krishnakumar Gopalakrishnan \\
                    & July 24, 2019\\
                    & London, United Kingdom \\
              \end{tabular}
    \end{flushright}
```

Look at the mis-aligned `\end{flushright}`. 

Can you please reopen this issue and consider a more robust long-term solution for this?

**Comment by andymass** [comment]
@krishnakumarg1984, can `latexindent.pl` operate on a partial file?  Or with a base indentation offset? 
 Using `equalprg` with many external tools is difficult because `equalprg` is called over a subset of lines.  It might be better to use `indentexpr` with `job_start`/`system` instead.

A similar problem exists with clang-format, since context information is essential for indentation in C++, the plugin https://github.com/rhysd/vim-clang-format came a solution where a map calls `%!clang-format -lines={visual selection}`.

**Comment by lervag** [comment]
@krishnakumarg1984 Yes, we'll continue this work. Please update, though, because I think the issue should not occur for your provided example...

I also would tend to agree that it could be nice to delegate the task to latexindent, if possible. I would be very happy if anyone is able to provide some proof of concept for me to work on.

**Comment by Konfekt** [comment]
Yes it's tricky, now
```tex
\begin{equation*}
	\xymatrix{
	  A\ar {\pi} & B
	}
\end{equation*}
```
is off by one. The braces around `\pi` cause this.

**Comment by krishnakumarg1984** [comment]
@lervag I have just now updated `vimtex`, quit & reopened `neovim-0.4.0-dev` and still have this issue on my example. Can you please test at your end with my example?

**Comment by krishnakumarg1984** [comment]
A full-blown latex-specific indentation algorithm implemented in Perl seems like the right way to do this.  

@andymass Not sure. I don't think we can pipe some text from `stdin` into `latexindent.pl`. From its documentaton, it looks like there is no check on whether the file is a valid document, so my suggestion would be to write the selected text to a temp file, run `latexindent.pl` with the `--overwrite` flag on it, which writes out the indented file. Then, `vimtex` should read the file back, replacing the shipped out lines in the buffer with the contents of the indented output from `latexindent`.

**Comment by andymass** [comment]
@krishnakumarg1984 This basically works, although it uses formatexpr instead of indentexpr.  If latexindent.pl makes/breaks lines it would be difficult to use as an indentexpr.

```vim
function! LatexFormat() abort
    let l:first = v:lnum
    let l:last = v:lnum + v:count - 1
    let l:lines = systemlist(
        \ 'latexindent 2>/dev/null',
        \ getline(l:first, l:last))
    if v:shell_error
        echoerr 'error:' v:shell_error
        return
    endif
    let l:offset = indent(l:first)
    let l:diff = len(l:lines) - (l:last - l:first + 1)
    if l:diff < 0
        call deletebufline('', l:last - l:diff + 1, l:last)
    elseif l:diff > 0
        call append(l:last, repeat([''], l:diff))
    endif
    call map(l:lines, 'repeat(" ", l:offset) . v:val')
    call setline(l:first, l:lines)
endfunction

set formatexpr=LatexFormat()
```

If on the other hand latexindent doesn't break lines, it would be possible to cache the result of the utility then pass it back from the indentexpr.

**Comment by krishnakumarg1984** [comment]
@andymass  thanks a lot. So we have to use gq instead of = . Got it.

Can you also update the code so that function is active only when latexindent is found on the PATH?
